### DISCUSSION TITLE
Dynamic Form CRUD

### DISCUSSION THREAD

**Discussion by Hallidayo** [prisma/prisma#11907] [Q&A] [answered]
Hi,

I currently have a dynamic form which lets users add and delete rows in a form for recipes but I am trying to figure out how I would update the data. Here is an example schema:

```prisma
model Recipes {
  id          String              @id @default(uuid())
  name        String
  method      RecipeMethods[]
  ingredients RecipeIngredients[]
  createdAt   DateTime            @default(now())
  updatedAt   DateTime            @updatedAt
}

model RecipeMethods {
  id        String   @id @default(uuid())
  method    String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  Recipes   Recipes? @relation(fields: [recipesId], references: [id])
  recipesId String?
}

model RecipeIngredients {
  id         String   @id @default(uuid())
  ingredient String?
  Recipes    Recipes? @relation(fields: [recipesId], references: [id])
  recipesId  String?
}
```

Say for example a user has the following record which has been created and they want to edit the record:

```ts
{
  "id": "1",
  "name": "Recipe Name",
  "method": [ {id:1, method: "method 1", recipesId:  1} , {id:2, method: "method 2", recipesId:  1}],
  "ingredients": [{id:1, ingredient: "ingredient 1", recipesId:  1} , {id:2, ingredient: "ingredient 2", recipesId:  1}]
}
```

They then make the following edits on the form:

```ts
{
  "id": "1",
  "name": "Recipe Name",
  "method": [ {id:1, method: "method 1 example", recipesId:  1} , {id:3, method: "method 3", recipesId:  1}],
  "ingredients": [{id:2, ingredient: "ingredient 2 change", recipesId:  1}]
}
```

So in the method array: 

1. `UPDATE` - The method of id 1 has been changed from "method 1" to "method 1 example"
2. `DELETE` - The method with id 2 has been deleted
3. `INSERT` - Method 3 has now been added.

The ingredients array:

1. `DELETE` - The first ingredient with id 1 has been deleted
2. `UPDATE` - The ingredient with id 2 has updated the ingredient from "ingredient 2" to "ingredient 2 change"

How would I go to achieve the above in Prisma?

Thankyou.

**Comment by austincrim** [comment, marked as answer]
Hey @Hallidayo 👋,

You can use our [nested writes API](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes) to accomplish this! It might look something like this:

```ts
const updated = await prisma.recipes.update({
  where: {
    id: "...",
  },
  data: {
    method: {
      update: {
        where: {
          id: "...",
        },
        data: {
          method: "method 1 example",
        },
      },
      delete: {
        id: "...",
      },
      create: {
        method: "method 3",
      },
    },
    ingredients: {
      delete: {
        id: "...",
      },
      update: {
        where: {
          id: "...",
        },
        data: {
          ingredient: "ingredient 2 change",
        },
      },
    },
  },
});
```

You will need to know the IDs and operations for each `ingredient` or `method`, so maybe you can send that info from the frontend. Otherwise, you could try to do some diffing between the new data and the old data to figure which operations you need to do.

Let me know if I can clarify anything!

**Reply by Hallidayo** [reply]
Hi @austincrim 

Thank you for your reply and pointing me in the right direction.

Have you got any examples or ideas on how I could compare the original data and edited data? I've been trying to work out what the best way to do this is. I was thinking of maybe doing a comparison of the data and pushing any inserts,updates or deletes into separate objects but I think this may be a bit too long winded?

**Reply by austincrim** [reply]
If you can't change the payload from the frontend, I think that might be the only way to do it!

**Reply by Hallidayo** [reply]
Hi @austincrim 

Again thank you for your help on this issue. 

I've created a function which does a comparison of the original data and new edited data on the form by the user. This function is now returning three arrays which are Create, Update and Delete. 

I am then using the nested writes with the return values of my function. 

All seems to be working well.

Thankyou.

**Comment by KayBeSee** [comment]
Regarding this question, do you have any general advice on managing the types of the create vs. update form component? 

Ideally, this would be the same form component, probably with a different `onSubmit` method passed in, depending on if it's being used in the update vs. create context. 

I am having trouble getting the types to play nicely with each other.

Using your recipe example above, I figure you have a form component that has a `recipe` prop and `setRecipe` prop passed into it:

```
import { Prisma } from "prisma";

interface Props {
  recipe: Prisma.RecipeGetPayload<{
    include: {
      method: true;
      ingredients: true;
    };
  }>;
  updateRecipe: (recipe: Prisma.RecipeCreateInput) => void;
  onSubmit: (recipe: Prisma.RecipeCreateInput | Prisma.RecipeUpdateInput) => void;
}

const RecipeForm = ({ recipe, updateRecipe, onSubmit }: Props) => {
  return (
    <form onSubmit={onSubmit}>
      <input
        value={recipe.name}
        onChange={(e) => {
          const updatedRecipe = { ...recipe };
          updatedRecipe.name = e.target.value;
          updateRecipe(updatedRecipe);
        }}
      />

      <button type="submit">Save Recipe</button>
    </form>
  );
};

export default RecipeForm;
```

The problem comes when I want to create relationships. I can't have a recipeId on the RecipeMethod if it isn't created already.

Any advice?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
